# Non-canonical start-codons and where to find them
Project about evolution of non-canonical start-codons in bacteria and connections between gene function and its start-codon.

## Short description

## Table of content

[Short description](https://github.com/Asklepiad/BI_project_2022/tree/main#short-description)

[Structure of repository](https://github.com/Asklepiad/BI_project_2022/tree/main#structure-of-repository)

[Introduction](https://github.com/Asklepiad/BI_project_2022/tree/main#introduction)

[Aims and objectives](https://github.com/Asklepiad/BI_project_2022/tree/main#aims-and-objectives)

[Data](https://github.com/Asklepiad/BI_project_2022/tree/main#data)

[Workflow](https://github.com/Asklepiad/BI_project_2022/tree/main#workflow)

[Workflow plan and technical properties](https://github.com/Asklepiad/BI_project_2022/tree/main#workflow-plan-and-technical-properties)

[Installation](https://github.com/Asklepiad/BI_project_2022/tree/main#installation)

[Options](https://github.com/Asklepiad/BI_project_2022/tree/main#options)

[Example input and output](https://github.com/Asklepiad/BI_project_2022/tree/main#example-input-and-output)

[Additional info](https://github.com/Asklepiad/BI_project_2022/tree/main#additional-info)

[Results and discussion](https://github.com/Asklepiad/BI_project_2022/tree/main#results-and-discussion)

[Future plans](https://github.com/Asklepiad/BI_project_2022/tree/main#future-plans)

[Literature](https://github.com/Asklepiad/BI_project_2022/tree/main#literature)

[Links](https://github.com/Asklepiad/BI_project_2022/tree/main#links)

[Authors](https://github.com/Asklepiad/BI_project_2022/tree/main#authors)

[Feedback](https://github.com/Asklepiad/BI_project_2022/tree/main#feedback)

## Structure of repository
> Data

All data for reproducing of our investigation and json file for test run are situated in `data` folder. You can read more detailed description of the folder content in `data` folder [readme]()  (*in development*).

> Scripts

Conda environment creator for pipeline, master script, all components of pipeline, some additional scripts and requirements file are located in `scripts` folder. You can read more detailed description of the folder content in `scripts` folder [readme]()  (*in development*).

> Test output

The test running on *Salinibacter ruber* results are situated in `test output` folder (*in development*). You can read more detailed description of the folder content in  `test output` folder [readme]()  (*in development*).

> Illustrations

The images for illustrating repositry's readme files are located in `illustrations` folder (*in development*). You can read more detailed description of the folder content in  `illustrations` folder [readme]()  (*in development*).

## Introduction

It is known procariotes have non only ATG as start-codon. There are about 10% genes in bacterial genome (more strict proportion depends on taxon) with non-canonical start-codons (NCSCs): GTG, TTG and some other in negligibly small proportion. There are not very much data about causes of existance and "gene preferencies" of NCSCs. The GTG and TTG are usually binds weaker with the ribosome relative to canonical ATG start-codone.

## Aims and objectives

#### Aims

1. Find the connection between function of genes and canonicity of its start-codone.
2. Evaluate distribution of bacterial NCSCs from different taxons and ecological niches, find the difference (if it exists) between different groups of bacteria by their function-start-codons patterns.
3. Create pipeline for analysing bacteria.
4. Run pipeline on many organisms from different taxons and ecological niches.

## Data

First stage of investigation was done on thirty-four bacterial species. Complete list of species is available on `data/all_species.txt` file. Species were selected by material of D. Nikolaeva article preprint about generalists and specialists (*article haven't published yet, therefore we can't post any reference or link*).

The json file with precomputed links is situated in `data/id_lists.json`. This file is a main part complete pipeline's input.

## Workflow

### Workflow plan and technical properties

### Installation

### Options

### Example input and output

### Additional info

## Results and discussion

## Future plans

Improve quality control and belief in finding position of start-codon.

## Literature

## Links

## Authors

 - Rubinova Valeria Sergeevna (Bioinformatics institute)
 - Sotnikov Bogdan Vladimirovich (Kyrgyz-Russian Slavic university, Bioinformatics institute)
 - Bochkareva Olga Olegovna* (Institute of Science and Technology Austria)

\* project supervisor 

## Feedback


