{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import json\n",
    "import re\n",
    "import argparse\n",
    "import shutil\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from Bio.Seq import Seq\n",
    "from itertools import chain\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_name = \"C_psittaci\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "qTr549eT0tml"
   },
   "outputs": [],
   "source": [
    "# Function for parsing assembly number\n",
    "def get_assembly_number(pre_pattern1, string):\n",
    "    pre_pattern2 = r\"_[\\d]+?\\.gbk\"\n",
    "    first_bound = re.search(pre_pattern1, string)\n",
    "    second_bound = re.search(pre_pattern2, string)\n",
    "    result = string[first_bound.end():(second_bound.start())]\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "yjqDsdd99_20"
   },
   "outputs": [],
   "source": [
    "# Uploading the data \n",
    "current_path = f\"../{folder_name}/data/{folder_name}_annotate/\" # Path directory\n",
    "pre_pattern1 = f\"{folder_name}\"\n",
    "gb_records= []  # Reasserting the gb_records list\n",
    "gbk_files = os.listdir(current_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "yjqDsdd99_20"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 47/47 [00:02<00:00, 22.78it/s]\n"
     ]
    }
   ],
   "source": [
    "for gbk in tqdm(gbk_files):\n",
    "    assembly_number = get_assembly_number(pre_pattern1, gbk)\n",
    "    link = f\"{current_path}{gbk}\"\n",
    "    gbk_object = SeqIO.read(link, 'genbank')\n",
    "    gb_records.append((gbk_object, assembly_number))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ya-4wM01v5OE"
   },
   "source": [
    "### Creating first table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uploading info about which DNA is plasmid\n",
    "with open(f\"../{folder_name}/data//{folder_name}_plasmid_code1.json\", \"r\") as tuples:\n",
    "    tuples = json.load(tuples)\n",
    "with open(f\"../{folder_name}/data/{folder_name}_plasmid_code2.json\", \"r\") as dna_type:\n",
    "    dna_type = json.load(dna_type)\n",
    "tuples = [tuple(x) for x in tuples]\n",
    "plasmid_code = dict(zip(tuples, dna_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "248o-4h8wHDZ"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 47/47 [00:09<00:00,  5.12it/s]\n"
     ]
    }
   ],
   "source": [
    "# Parsing\n",
    "locus_tag, start_codone, source, n_sequence, aa_sequence, DNA_source, \\\n",
    "product, cog, p_c_unity, first_n, last_n, strand_n, \\\n",
    "length_n = [], [], [], [], [], [], [], [], [], [], [], [], []\n",
    "for record_number in tqdm(range(len(gb_records))):\n",
    "    for feature_number in range(1, len(gb_records[record_number][0].features)):\n",
    "        if gb_records[record_number][0].features[feature_number].type == \"CDS\":\n",
    "            \n",
    "            ## Column 1 -- uniq id\n",
    "            if 'locus_tag' in gb_records[record_number][0].features[feature_number].qualifiers.keys():\n",
    "                locus_tag.append(gb_records[record_number][0].features[feature_number].qualifiers['locus_tag'][0])\n",
    "            else:\n",
    "                locus_tag.append(\"absent\")\n",
    "                \n",
    "            ## Column 2 -- source\n",
    "            source.append(gb_records[record_number][0].name)\n",
    "            \n",
    "            ## Column 3, 11-14 -- nucleotide sequence & 7 -- start-codone\n",
    "            first = gb_records[record_number][0].features[feature_number].location.start\n",
    "            last = gb_records[record_number][0].features[feature_number].location.end\n",
    "            strand = gb_records[record_number][0].features[feature_number].location.strand\n",
    "            first_n.append(first)\n",
    "            last_n.append(last)\n",
    "            strand_n.append(strand)\n",
    "            length_n.append(last - first)\n",
    "            if strand == 1:\n",
    "                sequence = str(gb_records[record_number][0].seq[first:last])\n",
    "                n_sequence.append(sequence)\n",
    "                start_codone.append(sequence[0:3])\n",
    "            elif strand == -1:\n",
    "                sequence = str(gb_records[record_number][0].seq.complement()[first:last])[::-1]\n",
    "                n_sequence.append(sequence)\n",
    "                start_codone.append(sequence[0:3])\n",
    "            else:\n",
    "                n_sequence.append(\"absent\")\n",
    "                \n",
    "            ## Column 4 -- aminoacid sequence\n",
    "            if 'translation' in gb_records[record_number][0].features[feature_number].qualifiers.keys():\n",
    "                aa_sequence.append(gb_records[record_number][0].features[feature_number].qualifiers['translation'][0])\n",
    "            else:\n",
    "                aa_sequence.append(\"absent\")\n",
    "                \n",
    "            ## Column 5 -- plasmid or chromosome\n",
    "            index = re.search(r\"[\\d]+?_[\\d]+$\", gb_records[record_number][0].name)\n",
    "            pc_key = tuple([int(i) for i in index.group().split(\"_\")])\n",
    "            if plasmid_code[pc_key] == \"plasmid\":\n",
    "                DNA_source.append(\"plasmid\")\n",
    "            else:\n",
    "                DNA_source.append(\"chromosome\")\n",
    "            \n",
    "            # Column 6 -- protein, coding by this gene\n",
    "            if 'product' in gb_records[record_number][0].features[feature_number].qualifiers.keys():\n",
    "                product.append(gb_records[record_number][0].features[feature_number].qualifiers['product'][0])\n",
    "            else:\n",
    "                product.append(\"absent\")\n",
    "                \n",
    "            ## Column 7 -- COG number\n",
    "            if 'db_xref' in gb_records[record_number][0].features[feature_number].qualifiers.keys():\n",
    "                cog.append(gb_records[record_number][0].features[feature_number].qualifiers[\"db_xref\"][0][4:])\n",
    "            else:\n",
    "                cog.append(\"absent\")\n",
    "            \n",
    "            ## Column 8 -- number of molecule (for identification and further plasmid marking)\n",
    "            p_c_unity.append(gb_records[record_number][1])\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "2e8OTrYR8_ho"
   },
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame(\n",
    "    {\n",
    "        \"id\": pd.Series(locus_tag),\n",
    "        \"source\": pd.Series(source),\n",
    "        \"n_sequence\": pd.Series(n_sequence),\n",
    "        \"aa_sequence\": pd.Series(aa_sequence),\n",
    "        \"type_of_DNA_source\": pd.Series(DNA_source),\n",
    "        \"start_codone\": pd.Series(start_codone),\n",
    "        \"product\": pd.Series(product),\n",
    "        \"cog\": pd.Series(cog),\n",
    "        \"p_c_unity\": pd.Series(p_c_unity),\n",
    "        \"first\": pd.Series(first_n),\n",
    "        \"last\": pd.Series(last_n),\n",
    "        \"strand\": pd.Series(strand_n),\n",
    "        \"length\": pd.Series(length_n),\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>source</th>\n",
       "      <th>n_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "      <th>type_of_DNA_source</th>\n",
       "      <th>start_codone</th>\n",
       "      <th>product</th>\n",
       "      <th>cog</th>\n",
       "      <th>p_c_unity</th>\n",
       "      <th>first</th>\n",
       "      <th>last</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>atg_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LEBECCED_00001</td>\n",
       "      <td>C_psittaci4_1</td>\n",
       "      <td>GTGGTAGACTTTGCAACTTTAGGTGGTAGACTTTTGGTTGAAAAAA...</td>\n",
       "      <td>MVDFATLGGRLLVEKNLDCARLVHYKILRLSMSGLSHHHKSRLFLT...</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>GTG</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>absent</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>1033</td>\n",
       "      <td>-1</td>\n",
       "      <td>1017</td>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LEBECCED_00002</td>\n",
       "      <td>C_psittaci4_1</td>\n",
       "      <td>GTGAGTAAATTAACTAAAGAAGCATCCGCTTTTTTTCAAAAGAATC...</td>\n",
       "      <td>MSKLTKEASAFFQKNQENTTKEFLKKEFAMDVFSVSLSDIEKEQIE...</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>GTG</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>absent</td>\n",
       "      <td>4</td>\n",
       "      <td>1161</td>\n",
       "      <td>1902</td>\n",
       "      <td>-1</td>\n",
       "      <td>741</td>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LEBECCED_00003</td>\n",
       "      <td>C_psittaci4_1</td>\n",
       "      <td>GTGAAAACGCTAGCATTTTGTTCTTTTAAAGGTGGTACTGGGAAAA...</td>\n",
       "      <td>MKTLAFCSFKGGTGKTTLSLNIGSNLAQVSRKKVLLVDLDPQANLT...</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>GTG</td>\n",
       "      <td>Sporulation initiation inhibitor protein Soj</td>\n",
       "      <td>COG1192</td>\n",
       "      <td>4</td>\n",
       "      <td>1898</td>\n",
       "      <td>2678</td>\n",
       "      <td>-1</td>\n",
       "      <td>780</td>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LEBECCED_00004</td>\n",
       "      <td>C_psittaci4_1</td>\n",
       "      <td>GTGGTTTTATATGAAAAAATCATACGTCTATTTACTGGAGAGATTA...</td>\n",
       "      <td>MVLYEKIIRLFTGEIMNDKQNMTNDFIKIVKDVEKDFPELDIKMKV...</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>GTG</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>absent</td>\n",
       "      <td>4</td>\n",
       "      <td>2708</td>\n",
       "      <td>3062</td>\n",
       "      <td>-1</td>\n",
       "      <td>354</td>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LEBECCED_00005</td>\n",
       "      <td>C_psittaci4_1</td>\n",
       "      <td>ATGGGTAATTCTGGTTTTTACTTAAATGACACCCAAAATTGTGTAT...</td>\n",
       "      <td>MGNSGFYLNDTQNCVFADNIKLGQMESPLQDQQLILGTKSTPTAAK...</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>ATG</td>\n",
       "      <td>Virulence plasmid protein pGP3-D</td>\n",
       "      <td>absent</td>\n",
       "      <td>4</td>\n",
       "      <td>3111</td>\n",
       "      <td>3906</td>\n",
       "      <td>-1</td>\n",
       "      <td>795</td>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27330</th>\n",
       "      <td>OCODBOIF_01040</td>\n",
       "      <td>C_psittaci11_1</td>\n",
       "      <td>ATGCTTACGGATAAAATTATATTATTTGTTACCGAAGACAGTAGTA...</td>\n",
       "      <td>MLTDKIILFVTEDSSISLQLKEFAQNVEYNIVVSSTLTDTPEADLI...</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ATG</td>\n",
       "      <td>Atypical response regulator protein ChxR</td>\n",
       "      <td>absent</td>\n",
       "      <td>11</td>\n",
       "      <td>1158901</td>\n",
       "      <td>1159615</td>\n",
       "      <td>1</td>\n",
       "      <td>714</td>\n",
       "      <td>846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27331</th>\n",
       "      <td>OCODBOIF_01041</td>\n",
       "      <td>C_psittaci11_1</td>\n",
       "      <td>ATGATACTTGCATCTGTTTTATTTTCCCCCGAAGATTTTTCTTTTC...</td>\n",
       "      <td>MILASVLFSPEDFSFPELITEAYYTWDILALIDRKLSTHVFSGIHG...</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ATG</td>\n",
       "      <td>Bifunctional protein GlmU</td>\n",
       "      <td>absent</td>\n",
       "      <td>11</td>\n",
       "      <td>1159674</td>\n",
       "      <td>1160289</td>\n",
       "      <td>1</td>\n",
       "      <td>615</td>\n",
       "      <td>846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27332</th>\n",
       "      <td>OCODBOIF_01042</td>\n",
       "      <td>C_psittaci11_1</td>\n",
       "      <td>ATGGTCGTTATGGATATTTTCGAAGCTTACCGCGTAATGATAGAAA...</td>\n",
       "      <td>MVVMDIFEAYRVMIETGIEDSLDEFGPKGFSIRAPVEYALTSGGKR...</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ATG</td>\n",
       "      <td>Farnesyl diphosphate synthase</td>\n",
       "      <td>absent</td>\n",
       "      <td>11</td>\n",
       "      <td>1160304</td>\n",
       "      <td>1161186</td>\n",
       "      <td>1</td>\n",
       "      <td>882</td>\n",
       "      <td>846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27333</th>\n",
       "      <td>OCODBOIF_01044</td>\n",
       "      <td>C_psittaci11_1</td>\n",
       "      <td>ATGAAAAACCTTATCGATAATAACTTAGTGAGATTTAAAAATATTT...</td>\n",
       "      <td>MKNLIDNNLVRFKNISKTKQGIFVNFKVRGEKGGASFTASIAVDIE...</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ATG</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>absent</td>\n",
       "      <td>11</td>\n",
       "      <td>1161285</td>\n",
       "      <td>1161540</td>\n",
       "      <td>-1</td>\n",
       "      <td>255</td>\n",
       "      <td>846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27334</th>\n",
       "      <td>OCODBOIF_01045</td>\n",
       "      <td>C_psittaci11_1</td>\n",
       "      <td>ATGTTGAGCAGAGATGATGAGTTTTCTTCAGAACAAAGGAAGAGTT...</td>\n",
       "      <td>MLSRDDEFSSEQRKSLSHFVTNLETNIFALKNLPEVVKGALFSKYS...</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>ATG</td>\n",
       "      <td>Flavin-dependent thymidylate synthase</td>\n",
       "      <td>absent</td>\n",
       "      <td>11</td>\n",
       "      <td>1161622</td>\n",
       "      <td>1163215</td>\n",
       "      <td>-1</td>\n",
       "      <td>1593</td>\n",
       "      <td>846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27335 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id          source  \\\n",
       "0      LEBECCED_00001   C_psittaci4_1   \n",
       "1      LEBECCED_00002   C_psittaci4_1   \n",
       "2      LEBECCED_00003   C_psittaci4_1   \n",
       "3      LEBECCED_00004   C_psittaci4_1   \n",
       "4      LEBECCED_00005   C_psittaci4_1   \n",
       "...               ...             ...   \n",
       "27330  OCODBOIF_01040  C_psittaci11_1   \n",
       "27331  OCODBOIF_01041  C_psittaci11_1   \n",
       "27332  OCODBOIF_01042  C_psittaci11_1   \n",
       "27333  OCODBOIF_01044  C_psittaci11_1   \n",
       "27334  OCODBOIF_01045  C_psittaci11_1   \n",
       "\n",
       "                                              n_sequence  \\\n",
       "0      GTGGTAGACTTTGCAACTTTAGGTGGTAGACTTTTGGTTGAAAAAA...   \n",
       "1      GTGAGTAAATTAACTAAAGAAGCATCCGCTTTTTTTCAAAAGAATC...   \n",
       "2      GTGAAAACGCTAGCATTTTGTTCTTTTAAAGGTGGTACTGGGAAAA...   \n",
       "3      GTGGTTTTATATGAAAAAATCATACGTCTATTTACTGGAGAGATTA...   \n",
       "4      ATGGGTAATTCTGGTTTTTACTTAAATGACACCCAAAATTGTGTAT...   \n",
       "...                                                  ...   \n",
       "27330  ATGCTTACGGATAAAATTATATTATTTGTTACCGAAGACAGTAGTA...   \n",
       "27331  ATGATACTTGCATCTGTTTTATTTTCCCCCGAAGATTTTTCTTTTC...   \n",
       "27332  ATGGTCGTTATGGATATTTTCGAAGCTTACCGCGTAATGATAGAAA...   \n",
       "27333  ATGAAAAACCTTATCGATAATAACTTAGTGAGATTTAAAAATATTT...   \n",
       "27334  ATGTTGAGCAGAGATGATGAGTTTTCTTCAGAACAAAGGAAGAGTT...   \n",
       "\n",
       "                                             aa_sequence type_of_DNA_source  \\\n",
       "0      MVDFATLGGRLLVEKNLDCARLVHYKILRLSMSGLSHHHKSRLFLT...            plasmid   \n",
       "1      MSKLTKEASAFFQKNQENTTKEFLKKEFAMDVFSVSLSDIEKEQIE...            plasmid   \n",
       "2      MKTLAFCSFKGGTGKTTLSLNIGSNLAQVSRKKVLLVDLDPQANLT...            plasmid   \n",
       "3      MVLYEKIIRLFTGEIMNDKQNMTNDFIKIVKDVEKDFPELDIKMKV...            plasmid   \n",
       "4      MGNSGFYLNDTQNCVFADNIKLGQMESPLQDQQLILGTKSTPTAAK...            plasmid   \n",
       "...                                                  ...                ...   \n",
       "27330  MLTDKIILFVTEDSSISLQLKEFAQNVEYNIVVSSTLTDTPEADLI...         chromosome   \n",
       "27331  MILASVLFSPEDFSFPELITEAYYTWDILALIDRKLSTHVFSGIHG...         chromosome   \n",
       "27332  MVVMDIFEAYRVMIETGIEDSLDEFGPKGFSIRAPVEYALTSGGKR...         chromosome   \n",
       "27333  MKNLIDNNLVRFKNISKTKQGIFVNFKVRGEKGGASFTASIAVDIE...         chromosome   \n",
       "27334  MLSRDDEFSSEQRKSLSHFVTNLETNIFALKNLPEVVKGALFSKYS...         chromosome   \n",
       "\n",
       "      start_codone                                       product      cog  \\\n",
       "0              GTG                          hypothetical protein   absent   \n",
       "1              GTG                          hypothetical protein   absent   \n",
       "2              GTG  Sporulation initiation inhibitor protein Soj  COG1192   \n",
       "3              GTG                          hypothetical protein   absent   \n",
       "4              ATG              Virulence plasmid protein pGP3-D   absent   \n",
       "...            ...                                           ...      ...   \n",
       "27330          ATG      Atypical response regulator protein ChxR   absent   \n",
       "27331          ATG                     Bifunctional protein GlmU   absent   \n",
       "27332          ATG                 Farnesyl diphosphate synthase   absent   \n",
       "27333          ATG                          hypothetical protein   absent   \n",
       "27334          ATG         Flavin-dependent thymidylate synthase   absent   \n",
       "\n",
       "      p_c_unity    first     last  strand  length  atg_number  \n",
       "0             4       16     1033      -1    1017         888  \n",
       "1             4     1161     1902      -1     741         888  \n",
       "2             4     1898     2678      -1     780         888  \n",
       "3             4     2708     3062      -1     354         888  \n",
       "4             4     3111     3906      -1     795         888  \n",
       "...         ...      ...      ...     ...     ...         ...  \n",
       "27330        11  1158901  1159615       1     714         846  \n",
       "27331        11  1159674  1160289       1     615         846  \n",
       "27332        11  1160304  1161186       1     882         846  \n",
       "27333        11  1161285  1161540      -1     255         846  \n",
       "27334        11  1161622  1163215      -1    1593         846  \n",
       "\n",
       "[27335 rows x 14 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_stata = pd.DataFrame(df1.groupby(\"p_c_unity\").apply(lambda x: x.query(\"start_codone == 'ATG'\").shape[0]), )\n",
    "assembly_stata.columns = [\"atg_number\"]\n",
    "df1 = df1.merge(assembly_stata, on=\"p_c_unity\")\n",
    "k = assembly_stata[\"atg_number\"].median()\n",
    "df1 = df1.query(\"atg_number > @k/2\")\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "pn69F4e1d3rQ"
   },
   "source": [
    "For recoding COGs from ids to categories I have downloaded tsvs with all categories and appropriate COGs and have created python dictionary.\n",
    "Firstly I have uploaded tsvs to folder on google drive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "0yBHPpptpWCx"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 26/26 [00:00<00:00, 453.53it/s]\n"
     ]
    }
   ],
   "source": [
    "# Creating cog dictionary\n",
    "cog_dict = {}\n",
    "current_path = \"../data/bigcog/\"\n",
    "cog_cats = os.listdir(current_path)\n",
    "for cog_cat in tqdm(cog_cats):\n",
    "    table = pd.read_csv(os.path.join(current_path, cog_cat), sep=\"\\t\")\n",
    "    if len(table.Cat):\n",
    "        cat = table.Cat.tolist()\n",
    "        cat_ful = list(map(lambda x: x.split(\" \"),cat))\n",
    "        cat_ful_single = list(chain.from_iterable(cat_ful))\n",
    "        key = max(cat_ful_single, key = cat_ful_single.count)\n",
    "        value = table.COG.tolist()\n",
    "        cog_dict[key] = value\n",
    "cog_dict[\"S\"].append(\"absent\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['absent',\n",
       " 'absent',\n",
       " 'COG1192',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0305',\n",
       " 'absent',\n",
       " 'COG0113',\n",
       " 'COG1726',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1448',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0648',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1137',\n",
       " 'absent',\n",
       " 'COG2877',\n",
       " 'absent',\n",
       " 'COG0564',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0187',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG3869',\n",
       " 'absent',\n",
       " 'COG0233',\n",
       " 'COG0528',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0396',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1475',\n",
       " 'COG1123',\n",
       " 'COG0444',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0126',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0587',\n",
       " 'COG2271',\n",
       " 'absent',\n",
       " 'COG0124',\n",
       " 'COG0173',\n",
       " 'absent',\n",
       " 'COG0219',\n",
       " 'COG0526',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0802',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0774',\n",
       " 'COG0764',\n",
       " 'COG1043',\n",
       " 'COG0223',\n",
       " 'absent',\n",
       " 'COG0087',\n",
       " 'absent',\n",
       " 'COG0089',\n",
       " 'COG0090',\n",
       " 'COG0185',\n",
       " 'COG0091',\n",
       " 'COG0092',\n",
       " 'COG0197',\n",
       " 'absent',\n",
       " 'COG0186',\n",
       " 'absent',\n",
       " 'COG0198',\n",
       " 'COG0094',\n",
       " 'COG0096',\n",
       " 'COG0097',\n",
       " 'COG0256',\n",
       " 'COG0098',\n",
       " 'absent',\n",
       " 'COG0201',\n",
       " 'COG0099',\n",
       " 'absent',\n",
       " 'COG0202',\n",
       " 'COG0203',\n",
       " 'COG0057',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0632',\n",
       " 'COG0105',\n",
       " 'absent',\n",
       " 'COG0445',\n",
       " 'COG0305',\n",
       " 'absent',\n",
       " 'COG3202',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0237',\n",
       " 'COG1158',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0448',\n",
       " 'absent',\n",
       " 'COG0742',\n",
       " 'absent',\n",
       " 'COG0276',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0072',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2204',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0637',\n",
       " 'COG0101',\n",
       " 'COG1211',\n",
       " 'COG1408',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0217',\n",
       " 'absent',\n",
       " 'COG0766',\n",
       " 'COG0018',\n",
       " 'COG0204',\n",
       " 'COG0283',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0608',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0008',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0048',\n",
       " 'COG0049',\n",
       " 'COG0480',\n",
       " 'COG0051',\n",
       " 'COG0369',\n",
       " 'COG0245',\n",
       " 'absent',\n",
       " 'COG0261',\n",
       " 'COG0211',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1121',\n",
       " 'absent',\n",
       " 'COG0352',\n",
       " 'COG2145',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1115',\n",
       " 'absent',\n",
       " 'COG1734',\n",
       " 'COG1327',\n",
       " 'COG0307',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1301',\n",
       " 'COG0508',\n",
       " 'COG0517',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0112',\n",
       " 'absent',\n",
       " 'COG0253',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1427',\n",
       " 'COG1060',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0319',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2094',\n",
       " 'COG0557',\n",
       " 'absent',\n",
       " 'COG0576',\n",
       " 'COG1420',\n",
       " 'COG0442',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0436',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0537',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2876',\n",
       " 'absent',\n",
       " 'COG0834',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2262',\n",
       " 'COG0621',\n",
       " 'absent',\n",
       " 'COG0296',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0154',\n",
       " 'COG0064',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1039',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0030',\n",
       " 'absent',\n",
       " 'COG1154',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0242',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0424',\n",
       " 'COG0488',\n",
       " 'absent',\n",
       " 'COG1234',\n",
       " 'absent',\n",
       " 'COG0466',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0484',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1080',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0181',\n",
       " 'COG1066',\n",
       " 'COG0571',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0777',\n",
       " 'absent',\n",
       " 'COG1299',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0095',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0509',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1347',\n",
       " 'COG2869',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0593',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0275',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0769',\n",
       " 'COG0860',\n",
       " 'COG0776',\n",
       " 'absent',\n",
       " 'COG0825',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2107',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0623',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0564',\n",
       " 'absent',\n",
       " 'COG1694',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0631',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2255',\n",
       " 'absent',\n",
       " 'COG1523',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0260',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0635',\n",
       " 'absent',\n",
       " 'COG0508',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1191',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0362',\n",
       " 'COG0481',\n",
       " 'absent',\n",
       " 'COG3202',\n",
       " 'absent',\n",
       " 'COG0803',\n",
       " 'COG1121',\n",
       " 'COG1108',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1195',\n",
       " 'COG0592',\n",
       " 'COG0691',\n",
       " 'COG1477',\n",
       " 'COG0190',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0043',\n",
       " 'COG0227',\n",
       " 'COG1640',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG4789',\n",
       " 'absent',\n",
       " 'COG0012',\n",
       " 'absent',\n",
       " 'COG0130',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0195',\n",
       " 'COG0539',\n",
       " 'COG0492',\n",
       " 'COG0736',\n",
       " 'absent',\n",
       " 'COG0682',\n",
       " 'absent',\n",
       " 'COG0593',\n",
       " 'absent',\n",
       " 'COG0058',\n",
       " 'COG0508',\n",
       " 'COG0022',\n",
       " 'COG1071',\n",
       " 'absent',\n",
       " 'COG1044',\n",
       " 'absent',\n",
       " 'COG4775',\n",
       " 'COG0353',\n",
       " 'COG0332',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0236',\n",
       " 'COG0664',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1301',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0531',\n",
       " 'COG1830',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2011',\n",
       " 'COG1464',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0187',\n",
       " 'COG0188',\n",
       " 'COG0125',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0163',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0621',\n",
       " 'COG0267',\n",
       " 'absent',\n",
       " 'COG1136',\n",
       " 'absent',\n",
       " 'COG0102',\n",
       " 'COG0103',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0364',\n",
       " 'COG4166',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0765',\n",
       " 'COG1126',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0231',\n",
       " 'absent',\n",
       " 'COG0439',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0518',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG3842',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1212',\n",
       " 'COG0504',\n",
       " 'COG0816',\n",
       " 'COG4166',\n",
       " 'COG4166',\n",
       " 'COG4166',\n",
       " 'COG0601',\n",
       " 'COG1173',\n",
       " 'COG0444',\n",
       " 'COG0444',\n",
       " 'absent',\n",
       " 'COG0471',\n",
       " 'COG0205',\n",
       " 'absent',\n",
       " 'COG0205',\n",
       " 'COG1519',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1118',\n",
       " 'COG0767',\n",
       " 'COG0001',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0120',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0327',\n",
       " 'absent',\n",
       " 'COG0234',\n",
       " 'COG0459',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0507',\n",
       " 'COG0073',\n",
       " 'absent',\n",
       " 'COG0194',\n",
       " 'COG0164',\n",
       " 'COG0335',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0541',\n",
       " 'absent',\n",
       " 'COG0216',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0060',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1294',\n",
       " 'COG1271',\n",
       " 'COG1624',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0085',\n",
       " 'absent',\n",
       " 'COG0244',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0250',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0361',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1268',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0329',\n",
       " 'COG0527',\n",
       " 'absent',\n",
       " 'COG0289',\n",
       " 'COG3186',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1028',\n",
       " 'COG0502',\n",
       " 'COG0156',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0128',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0337',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1945',\n",
       " 'COG0531',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0039',\n",
       " 'absent',\n",
       " 'COG0166',\n",
       " 'COG4974',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0761',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0659',\n",
       " 'COG0114',\n",
       " 'absent',\n",
       " 'COG0563',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0024',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0590',\n",
       " 'absent',\n",
       " 'COG1185',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0016',\n",
       " 'absent',\n",
       " 'COG0291',\n",
       " 'COG0290',\n",
       " 'absent',\n",
       " 'COG0812',\n",
       " 'absent',\n",
       " 'COG0208',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0074',\n",
       " 'COG0045',\n",
       " 'COG0552',\n",
       " 'COG0670',\n",
       " 'COG0814',\n",
       " 'COG0449',\n",
       " 'COG1109',\n",
       " 'COG0617',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0333',\n",
       " 'absent',\n",
       " 'COG1530',\n",
       " 'absent',\n",
       " 'COG1025',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1947',\n",
       " 'COG0359',\n",
       " 'absent',\n",
       " 'COG0360',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0297',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0249',\n",
       " 'COG0322',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0199',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0215',\n",
       " 'absent',\n",
       " 'COG0526',\n",
       " 'absent',\n",
       " 'COG1198',\n",
       " 'absent',\n",
       " 'COG0204',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0221',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1060',\n",
       " 'COG0324',\n",
       " 'COG1366',\n",
       " 'absent',\n",
       " 'COG1324',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0707',\n",
       " 'COG0772',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0472',\n",
       " 'COG0770',\n",
       " 'COG0459',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0231',\n",
       " 'COG0775',\n",
       " 'absent',\n",
       " 'COG0013',\n",
       " 'COG1197',\n",
       " 'COG0407',\n",
       " 'absent',\n",
       " 'COG1232',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG2265',\n",
       " 'absent',\n",
       " 'COG2871',\n",
       " 'absent',\n",
       " 'COG1674',\n",
       " 'COG1235',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1115',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0054',\n",
       " 'COG0108',\n",
       " 'COG0117',\n",
       " 'COG0172',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0772',\n",
       " 'COG0340',\n",
       " 'absent',\n",
       " 'COG1187',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1157',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1274',\n",
       " 'COG1077',\n",
       " 'absent',\n",
       " 'COG0544',\n",
       " 'COG0740',\n",
       " 'COG1219',\n",
       " 'absent',\n",
       " 'COG1160',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0486',\n",
       " 'absent',\n",
       " 'COG1114',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG4791',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG1459',\n",
       " 'COG2804',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0006',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0441',\n",
       " 'COG1192',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0180',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0479',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0811',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0450',\n",
       " 'COG0459',\n",
       " 'absent',\n",
       " 'COG0127',\n",
       " 'absent',\n",
       " 'COG0692',\n",
       " 'COG0210',\n",
       " 'COG1508',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0042',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0468',\n",
       " 'COG5424',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0568',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0593',\n",
       " 'absent',\n",
       " 'COG0058',\n",
       " 'absent',\n",
       " 'absent',\n",
       " 'COG0022',\n",
       " ...]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "J2LzibiOtE-q"
   },
   "source": [
    "On the next stage I have created 25 new columns for dataset, each contained data if protein has this COG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "id": "G586jo7svBP8"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:01<00:00, 23.28it/s]\n"
     ]
    }
   ],
   "source": [
    "# Adding 25 columns to dataset\n",
    "for key in tqdm(cog_dict.keys()):   # Passing over COG categories\n",
    "    cog_list = []             # For each category create list\n",
    "    for cog_id in cog:        # Passing out every item\n",
    "        if cog_id not in cog_dict[key]:     # For every gene, if gene's id not in dict\n",
    "            cog_list.append(0)              # Mark it\n",
    "        else:                               \n",
    "            cog_list.append(1)              # Else mark another\n",
    "    df1[key] = pd.Series(cog_list)   # Creating 25 new columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3098    1\n",
       "Name: S, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.query(\"id == 'APHAOMEK_00005'\")[\"S\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_csv(f\"../{folder_name}/data/First_table.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 27/27 [00:00<00:00, 621.60it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n",
      "26\n",
      "6\n",
      "17\n",
      "12\n",
      "0\n",
      "3\n",
      "16\n",
      "10\n",
      "21\n",
      "7\n",
      "20\n",
      "19\n",
      "13\n",
      "2\n",
      "15\n",
      "1\n",
      "22\n",
      "18\n",
      "23\n",
      "24\n",
      "4\n",
      "14\n",
      "25\n",
      "5\n",
      "11\n",
      "8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "for source in tqdm(set(df1[\"p_c_unity\"])):\n",
    "    subset = df1[df1[\"p_c_unity\"] == source]\n",
    "    with open (f\"../{folder_name}/data/orto_rows/{folder_name}{str(source)}.fasta\", \"w\") as protein_fasta:\n",
    "        for index, row in subset.iterrows():\n",
    "            #if row['type_of_the_gene'] != \"pseudogene\":   # It will be actual, if we will found pseudogenes again\n",
    "            protein_fasta.write(\">\")\n",
    "            protein_fasta.write(row[\"id\"])\n",
    "            protein_fasta.write(\"_\")\n",
    "            protein_fasta.write(row[\"source\"])\n",
    "            protein_fasta.write(\"\\n\")\n",
    "            protein_fasta.write(row[\"aa_sequence\"])\n",
    "            protein_fasta.write(\"\\n\")"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [
    "-WDRF29W3SyR",
    "c8yV-DwPre6C",
    "kyeKwO01rj1G",
    "Zyu27Lga0wWR",
    "dc_xdF-VxoMY",
    "2vlUfrmdJmyr",
    "qlF4YG3srZrF",
    "uuQtazSJwK_s",
    "N3J1A7ThS3Xz",
    "Kk6luKhoS3X0",
    "XrODV9lQS3X0"
   ],
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
